jina/peapods/peas/dynamic_head_pea.py

```diff
@@ -0,0 +1,134 @@
+import copy
+import threading
+import time
+import zmq
+
+
+def get_next_targets(routes):
+    next_routes = {}
+    next_targets = routes[0]
+    if not isinstance(next_targets, list):
+        next_targets = [next_targets]
+    remaining_routes = routes[1:]
+    for next_target in next_targets:
+        if isinstance(next_target, list):
+            next_pod = next_target[0]
+            next_routes[next_pod] = next_target[1:] + remaining_routes
+        else:
+            next_routes[next_target] = remaining_routes
+    return next_routes
+
+
+class HeadPea:
+    def __init__(self, in_socket):
+        self._in_socket = in_socket
+        self.context = zmq.Context()
+        self.routes = {}
+        self.receiver = self.context.socket(zmq.ROUTER)
+        self.receiver.bind(self._in_socket)
+
+    def run(self):
+        while True:
+            self.receiver.recv()
+            message = self.receiver.recv_json()
+
+            self.do_work(message)
+            self.post(message)
```

Comment:
if i understand it correctly, these lines here basically summarizes head pea, worker pea and tail pea into one, for demo purpose.

Roughly head pea:
```python
self.receiver.recv()
message = self.receiver.recv_json()
```

Roughly worker pea
```python
self.do_work(message)
```

Roughly tail pea:
```python
self.post(message)
```

Am I right?


This code is for demo purpose, but in practice, the following lines should be implemented with `ZmqStreamlet` or `AsyncZmqlet` as we currently use them in ZEDRuntime:
```
self.receiver.recv()
...
self.post(message)
```

Reply:
am I also right about the affection here?
![image](https://user-images.githubusercontent.com/2041322/120729363-8ddb2f80-c511-11eb-84d9-da2712a079bf.png)

Reply:
Yes, all of you observation are right. Only the following three communication types are affected:

- Gateway -> HeadPea
- TailPea->HeadPea 
- TailPea -> Gateway

If we treat Gateway as a combination of a HeadPea and a TailPea (from a networking perspective) it boils down to `TailPea -> HeadPea`.

Reply:
So talking in current `naming` it seems that every HeadPea will be a `ROUTER server (bind) (ROUTER_BIND)` and every TailPea a `DEALER client (connect) (DEALER_CONNECT)`, right?

Reply:
https://stackoverflow.com/questions/19070140/zeromq-dealer-client-to-multiple-servers-router

this may be good to take into account